examples/gno.land/p/demo/grc/grc721/basic_nft_test.gno

```diff
@@ -135,6 +135,41 @@ func TestOwnerOf(t *testing.T) {
 	}
 }
 
+func TestSetTokenURI(t *testing.T) {
+	dummy := NewBasicNFT(dummyNFTName, dummyNFTSymbol)
+	if dummy == nil {
+		t.Errorf("should not be nil")
+	}
+
+	addr1 := users.AddressOrName("g1var589z07ppjsjd24ukm4uguzwdt0tw7g47cgm")
+	tokenURI := "http://example.com/token"
+
+	dummy.mint(addr1.Resolve(), TokenID("1"))
+	dummy.SetTokenURI(TokenID("1"), TokenURI(tokenURI))
+	
+	// Test case: Invalid token ID
+	_, err := dummy.SetTokenURI(TokenID("2"), TokenURI(tokenURI))
+	if err != ErrInvalidTokenId {
+		t.Errorf("Expected error %v, got %v", ErrInvalidTokenId, err)
+	}
+
+	// Test case: Caller is not owner
+	owner, err := dummy.OwnerOf(TokenID("1"))
```

Comment:
Don't see SetTokenURI being called here.

You can modify the caller by using `std.TestSetOrigCaller`

Reply:
Thanks for reviewing my PR, I updated the code with your suggestions. I have one outstanding issue which I can't really understand. I refactored the test case for "caller is not owner", it now looks like this:

```
std.TestSetOrigCaller(std.Address("g1us8428u2a5satrlxzagqqa5m6vmuze025anjlj")) // addr2
	
_, cerr := dummy.SetTokenURI(TokenID("1"), TokenURI(tokenURI))  // addr2 trying to set URI for token 1
if cerr != nil || cerr != ErrCallerIsNotOwner {
	t.Errorf("Expected error %v for non-owner setting token URI, got %v", ErrCallerIsNotOwner, cerr)
}
```

As far as I understand:

- `std.TestSetOrigCaller()` makes the next call to `SetTokenURI` to come from addr2
- since addr2 didn't mint token id 1, the call should fail with the error type: `ErrCallerIsNotOwner`

Instead, I get this weird error (sorry for pasting the entire output):

```
panic: reflect: reflect.Value.SetString using value obtained using unexported field

goroutine 1 [running]:
reflect.flag.mustBeAssignableSlow(0x14000da2af8?)
        /usr/local/go/src/reflect/value.go:259 +0xb4
reflect.flag.mustBeAssignable(...)
        /usr/local/go/src/reflect/value.go:249
reflect.Value.SetString({0x104babf80?, 0x140042c5240?, 0x14000da2b88?}, {0x140004b8ac1, 0x19})
        /usr/local/go/src/reflect/value.go:2437 +0x44
github.com/gnolang/gno/gnovm/pkg/gnolang.gno2GoValue(0x140004266f0, {0x104babf80?, 0x140042c5240?, 0x104244850?})
        /Users/dragosroua/gno/gnovm/pkg/gnolang/gonative.go:1091 +0x2d0
github.com/gnolang/gno/gnovm/pkg/gnolang.gno2GoValue(0x14000426720, {0x14008095f20?, 0x140042c5240?, 0x10423c774?})
        /Users/dragosroua/gno/gnovm/pkg/gnolang/gonative.go:1194 +0xc60
github.com/gnolang/gno/gnovm/pkg/gnolang.gno2GoValue(0x14004b0eff0, {0x104bcd320?, 0x14000544480?, 0x104b9ee20?})
        /Users/dragosroua/gno/gnovm/pkg/gnolang/gonative.go:1129 +0x548
github.com/gnolang/gno/gnovm/pkg/gnolang.gno2GoValue(0x14004b0f068, {0x104b9ee20?, 0x14006e3a648?, 0x140001bc168?})
        /Users/dragosroua/gno/gnovm/pkg/gnolang/gonative.go:1175 +0xd0c
github.com/gnolang/gno/gnovm/pkg/gnolang.(*Machine).doOpCallGoNative(0x140000d6b40)
        /Users/dragosroua/gno/gnovm/pkg/gnolang/gonative.go:1428 +0x220
github.com/gnolang/gno/gnovm/pkg/gnolang.(*Machine).Run(0x140000d6b40)
        /Users/dragosroua/gno/gnovm/pkg/gnolang/machine.go:1219 +0xbcc
github.com/gnolang/gno/gnovm/pkg/gnolang.(*Machine).Eval(0x140000d6b40, {0x104cc74c8, 0x14008095e60})
        /Users/dragosroua/gno/gnovm/pkg/gnolang/machine.go:615 +0x478
main.runTestFiles(0x140000d6b40, 0x1400000dda0, {0x1400032c318, 0x6}, 0x0, 0x0, {0x0, 0x0}, 0x0?)
        /Users/dragosroua/gno/gnovm/cmd/gno/test.go:402 +0x348
main.gnoTestPkg({0x140000381a7, 0x3}, {0x14000330400?, 0x1, 0x1048f5379?}, {0x0, 0x0, 0x1048446bc?}, 0x1400004a3c0, 0x1400029c5f0)
        /Users/dragosroua/gno/gnovm/cmd/gno/test.go:296 +0x348
main.execTest(0x1400004a3c0, {0x140003302d0, 0x1, 0x1}, 0x0?)
        /Users/dragosroua/gno/gnovm/cmd/gno/test.go:208 +0x7f0
main.newTestCmd.func1({0x0?, 0x0?}, {0x140003302d0?, 0x14000332b98?, 0x0?})
        /Users/dragosroua/gno/gnovm/cmd/gno/test.go:51 +0x38
github.com/gnolang/gno/tm2/pkg/commands.(*Command).Run(0x0?, {0x104cc1168?, 0x140000ae010?})
        /Users/dragosroua/gno/tm2/pkg/commands/command.go:233 +0x17c
github.com/gnolang/gno/tm2/pkg/commands.(*Command).Run(0x14000332840?, {0x104cc1168?, 0x140000ae010?})
        /Users/dragosroua/gno/tm2/pkg/commands/command.go:237 +0x12c
github.com/gnolang/gno/tm2/pkg/commands.(*Command).ParseAndRun(0x1400009e000?, {0x104cc1168, 0x140000ae010}, {0x140000ac190?, 0x60?, 0x0?})
        /Users/dragosroua/gno/tm2/pkg/commands/command.go:118 +0x4c
main.main()
        /Users/dragosroua/gno/gnovm/cmd/gno/main.go:14 +0x74
        ```
        
        I have a culprit in mind, namely the `owners` private property of the NFT, but I can't really wrap my head around how this SetString call fails in this case, but in the one above doesn't. Any pointers?

Reply:
nvm the abive, it was about smth else. Fixed the tests and the implementation, should be fine now.